src/rust/catmem/queue.rs

```diff
@@ -5,68 +5,292 @@
+        }
+    }
+
+    /// This function creates a new CatmemQueue and attaches to an existing share ring buffer as either a consumer or
+    /// producer as indicated by [mode].
+    pub fn open(name: &str, mode: QMode) -> Result<Self, Fail> {
+        match mode {
+            QMode::Push => Ok(Self {
+                ring: Rc::new(RefCell::<Ring>::new(Ring::open_push_ring(name)?)),
+                pending_ops: Rc::new(RefCell::<HashMap<TaskHandle, YielderHandle>>::new(HashMap::<
+                    TaskHandle,
+                    YielderHandle,
+                >::new(
+                ))),
+            }),
+            QMode::Pop => Ok(Self {
+                ring: Rc::new(RefCell::<Ring>::new(Ring::open_pop_ring(name)?)),
+                pending_ops: Rc::new(RefCell::<HashMap<TaskHandle, YielderHandle>>::new(HashMap::<
+                    TaskHandle,
+                    YielderHandle,
+                >::new(
+                ))),
+            }),
+        }
+    }
+
+    /// This function closes a ring endpoint.
+    /// TODO merge this with async_close().
+    pub fn close(&self) -> Result<(), Fail> {
+        match &mut *self.ring.borrow_mut() {
+            Ring::PushOnly(ring) => {
+                // Attempt to push EoF.
+                // Maximum number of retries. This is set to an arbitrary small value.
+                for _ in 0..MAX_RETRIES_PUSH_EOF {
+                    match ring.try_close() {
+                        Ok(()) => return Ok(()),
+                        Err(_) => continue,
+                    }
+                }
+                let cause: String = format!("failed to push EoF");
+                error!("push_eof(): {}", cause);
+                Err(Fail::new(libc::EIO, &cause))
+            },
+            // Nothing to do on close for a pop ring.
+            Ring::PopOnly(_) => Ok(()),
+        }
+    }
+
+    /// Prepares a transition to the closing state.
+    pub fn prepare_close(&self) -> Result<(), Fail> {
+        match &mut *self.ring.borrow_mut() {
+            Ring::PushOnly(ring) => ring.prepare_close(),
+            Ring::PopOnly(ring) => ring.prepare_close(),
         }
     }
 
-    /// Get underlying uni-directional pipe.
-    pub fn get_pipe(&self) -> &Pipe {
-        &self.pipe
+    /// Prepares a transition to the closed state.
+    pub fn prepare_closed(&self) -> Result<(), Fail> {
+        match &mut *self.ring.borrow_mut() {
+            Ring::PushOnly(ring) => ring.prepare_closed(),
+            Ring::PopOnly(ring) => ring.prepare_closed(),
+        }
     }
 
-    /// Set underlying uni-directional pipe.
-    pub fn get_mut_pipe(&mut self) -> &mut Pipe {
-        &mut self.pipe
+    /// This function commits the queue to closing.
+    pub fn commit(&self) {
+        match &mut *self.ring.borrow_mut() {
+            Ring::PushOnly(ring) => ring.commit(),
+            Ring::PopOnly(ring) => ring.commit(),
+        }
+    }
+
+    /// This function aborts a close operation.
```

Comment:
Does this apply to a close operation only? Or can `abort()` be called at anytime (even if not closing)?

Reply:
Good point, will update.

Reply:
`abort` may be called at any time. Comment was update to reflect this.